### PR TITLE
Runtime diagnostics for leaked messages in unbounded channels (part 2)

### PR DESCRIPTION
**Pull request by dmitry-markin**
This is a follow-up PR to https://github.com/paritytech/substrate/pull/12971 addressing code review issues.

polkadot companion: https://github.com/paritytech/polkadot/pull/6481

### CODE REVIEW

File: client/network/src/service/out_events.rs [hunk 1]
```diff
@@ -62,7 +62,7 @@ pub fn channel(name: &'static str, queue_size_warning: i64) -> (Sender, Receiver
 		queue_size: queue_size.clone(),
 		queue_size_warning,
 		warning_fired: false,
-		creation_backtrace: Backtrace::capture(),
+		creation_backtrace: Backtrace::new_unresolved(),
```
Comment (bkchr) [inline client/network/src/service/out_events.rs:65]: Can we not use the channel from utils here?

The metrics field is always None or?

File: client/utils/src/mpsc.rs [hunk 2]
```diff
@@ -149,23 +149,20 @@ mod inner {
...
-							"The number of unprocessed messages in channel `{}` reached {}.\n\
-							 The channel was created at:\n{}",
-							self.name, self.queue_size_warning, self.creation_backtrace,
-						),
-						_ => error!(
-							"The number of unprocessed messages in channel `{}` reached {}.",
-							self.name, self.queue_size_warning,
-						),
-					}
+					let mut bt = self.creation_backtrace.lock().expect("another thread panicked.");
```
Comment (bkchr) [inline client/utils/src/mpsc.rs:159]: I'm not really sure about the mutex here. I mean I get why, but we could also just use Arc and then clone the backtrace here into some mutable value to resolve it.

```
let mut backtrace = (*self.creation_backtrace).clone();
backtrace.resolve();
```

But I don't know if not using a mutex is wort the clone 🤷

File: client/network/src/service/out_events.rs (hunk 1 above)
Comment (dmitry-markin) [reply client/network/src/service/out_events.rs:65]: `Sender` is not accessible on its own, but only via `OutChannels`. `metrics` is inserted when a new `Sender` is added with `OutChannels::push()`:
https://github.com/paritytech/substrate/blob/80082f154829e3f29e1bb5b9d94c1307d7626b8d/client/network/src/service/out_events.rs#L176-L180

Comment (dmitry-markin) [reply client/network/src/service/out_events.rs:65]: I've updated the comment to indicate that `metrics` will be initialized.

Comment (dmitry-markin) [reply client/network/src/service/out_events.rs:65]: Metrics in `out_events.rs` and `mpsc.rs` seem different, so I don't know how to reuse `utils` version of the channel.

File: client/utils/src/mpsc.rs (hunk 2 above)
Comment (dmitry-markin) [reply client/utils/src/mpsc.rs]: Done. It's a rare one-time event anyway, not a big deal.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2198634

**Review by altonen — APPROVED** [review]

File: client/network/src/service/out_events.rs (hunk 1 above)
Comment (bkchr) [reply client/network/src/service/out_events.rs:65]: Okay ty. At some point we should probably merge both implementations. The metrics are probably not that different.

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**